app/models/concerns/notifications/membership_notifier.rb

```diff
@@ -0,0 +1,56 @@
+module Notifications
+  module MembershipNotifier
+    extend ActiveSupport::Concern
+
+    LEADER_KEYS = %w[membership.create].freeze
+    USER_KEYS   = %w[membership.accept membership.archive membership.inactivate].freeze
+
+    included do
+      acts_as_notifiable :users, targets: :targets
+      after_create :notify_from_create
+
+      after_update :notify_from_inactivate, if: :end_date_changed?
+      after_update :notify_from_archive, if: :archived_changed?
+    end
+
+    Membership.prepend(Module.new do
```

Comment:
Yeah, it is a good way to create magic..

but this is a tiny-shiny path that I've chosen to introduce this feature and later on this whole `Concern` can be changed after refactoring this state machine with `aasm` or something like that.